uat/testing-features/src/main/java/com/aws/greengrass/iot/IotCoreClient.java

```diff
@@ -0,0 +1,293 @@
+                    return;
+                } catch (ExecutionException | TimeoutException | IOException e) {
+                    log.error("unable to resubscribe to topic {}", request.getTopic(), e);
+                }
+            }
+        });
+    }
+
+    public void subscribe(SubscribeRequest request)
+            throws ExecutionException, InterruptedException, TimeoutException, IOException {
+        start();
+
+        SubscribePacket subscribePacket = new SubscribePacket.SubscribePacketBuilder()
+                .withSubscription(request.getTopic(), request.getQos())
+                .build();
+        Subscription subscription = Subscription.builder()
+                .topic(request.getTopic())
+                .callback(request.getCallback())
+                .request(request)
+                .build();
+        boolean newSubscription = subscriptions.add(subscription);
+
+        try {
+            SubAckPacket subAckPacket = client.subscribe(subscribePacket).get();
+            log.debug("subscription result: topic={}; reason={}; reasonCode={};",
+                    request.getTopic(),
+                    subAckPacket.getReasonString(),
+                    subAckPacket.getReasonCodes().get(0));
+            if (subAckPacket.getReasonCodes().get(0).getValue() > 2) {
+                if (newSubscription) {
+                    // only remove if not resubscribing
+                    subscriptions.remove(subscription);
+                }
+                request.getErrorCallback().accept(subAckPacket, null);
+                throw new RuntimeException("Unable to subscribe reasonCode="
+                        + subAckPacket.getReasonCodes().get(0));
+            }
+        } catch (ExecutionException | InterruptedException e) {
+            if (newSubscription) {
+                // only remove if not resubscribing
+                subscriptions.remove(subscription);
+            }
+            request.getErrorCallback().accept(null, e);
+            throw e;
+        }
+    }
+
+    private synchronized void start() throws InterruptedException, TimeoutException {
+        if (client == null) {
+            client = createMqttClient();
+            client.start();
+            log.info("Waiting for client to connect...");
+            if (!started.await(10L, TimeUnit.SECONDS)) {
+                throw new TimeoutException("Timed-out waiting for client connect on start");
+            }
+        }
+    }
+
+    public synchronized void close() {
+        if (client != null) {
+            try {
+                client.stop(new DisconnectPacket.DisconnectPacketBuilder()
+                        .withReasonCode(DisconnectPacket.DisconnectReasonCode.NORMAL_DISCONNECTION)
+                        .build());
+            } catch (CrtRuntimeException e) {
+                log.error("Unable to stop mqtt client", e);
+                client.close();
+            }
+        }
+        executorService.shutdownNow();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Call `shutdown()` to reject any incoming tasks and `awaitTermination()` to wait for existing tasks to complete or until a timeout is reached. To stop the execution and processing of tasks, if necessary, call `shutdownNow()`. Check the return value of `awaitTermination()` and handle when the method times out while waiting for other threads to stop executing. See the following diffs on GitHub that make this code change: [diff1](https://github.com/glowroot/glowroot/commit/81de5a3699c59e0cf98d485c380c1034e6cd28a5#diff-c608d2c40790b0e0e5228e6647f69729caa756bbe9127169139406314b80897eL143), [diff2](https://github.com/apache/directory-kerby/commit/3ed0e7cd3b8a0c3ce29293153e4e72f422f0637f#diff-eb42f072849f520563ec2a686789a836195f18075cc222ef355c6ea710b0a360L84)

Learn more: [ExecutorService in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html)